text/1010-global-vars.rst

```diff
@@ -0,0 +1,237 @@
+
+Required capabilities: DDL.
+
+Synopsis::
+
+    [ WITH <with-item> [, ...] ]
+    DROP GLOBAL <name> ;
+
+
+SET GLOBAL
+----------
+
+Set the value of a non-computed global variable *in the current session*
+by evaluating the given expression.
+
+Required capabilities: session config.
+
+Synopsis::
+
+    SET GLOBAL <name> := <expr> ;
+
+
+RESET GLOBAL
+------------
+
+Reset a non-computed global variable to its default value.
+
+Required capabilities: session config.
+
+Synopsis::
+
+    RESET GLOBAL <name> ;
+
+
+Referring to globals in queries
+===============================
+
+The new ``GLOBAL <name>`` expression is used to refer to the value of the
+given global variable in queries.  For example::
+
+    SELECT User FILTER .id = GLOBAL user_id
+
+The precedence of ``GLOBAL`` is similar to ``DETACHED`` and binds higher
+than most other operators, including the dot.
+
+References to globals are legal in regular EdgeQL queries (i.e. non-DDL) and
+in parts of schema definition that interpolate into EdgeQL queries:
+
+- defaults
+- expressions in computed properties and links
+- expression aliases
+
+
+Introspection
+=============
+
+Globals can be introspected via a new entry in the introspection schema:
+``schema::Global``.  Likewise a ``DESCRIBE GLOBAL <foo>`` statement returns
+a textual DDL or SDL description of a global.
+
+
+Performance and scalability considerations
+==========================================
+
+For scalability and HA reasons, the server must not maintain persistent session
+state, including the values of globals.  This means that it is the
+responsibility of the _client_ to maintain this state and send it to the server
+with every query or script request in a context header.  See
+`Protocol changes`_ below.
+
+Internally, references to non-computed globals are compiled as query arguments
+and are query cache friendly.
+
+
+Protocol changes
+================
+
+Regarding the protocol, there are two things to consider: client informing the
+server about the session state, and the server informing the client about
+the session state changes as a result of a ``SET/RESET GLOBAL`` statement.
+
+When sending a query or a script, a client _may_ specify the values of globals
+in the new "globals" header, encoded as a free object.
+
+When the server encounters a ``SET/RESET GLOBAL`` statement in a query or
+script message it *must* include the new values of the affected globals in
+the new "globals" header returned with the next ``CommandComplete`` response.
+
+
+Computed globals vs aliases
+===========================
+
+While there is a lot of similarity between computed globals and aliases,
+they aren't the same and complement each other.  Where an alias usually
+stands in place of a material set of objects (and this masquerades as an
+object type), a computed global stands in for a settable global.  Additionally,
+the alias cardinality is usually ``multi``, whereas most globals would normally
+be ``single``.
```

Comment:
We *could* skip having computed globals, right, and just use aliases? The *main* difference would be whether you write `global` when referring to it?

Reply:
Well, yes.  I think the distinction is important enough, though.

Reply:
The distinction is essentially just a question of documentation of intended purpose, though, right? To be honest it almost seems to me that we almost want aliases to just be subsumed by computed globals, though it's probably too late for that.

Reply:
Yes, it's a question of being able to identify references to global state in queries.  The primary intended purpose of aliases is to masquerade as schema types, i.e. they look and behave as if they were persistent sets (mostly).  Computed globals, on the other hand, masquerade as _global variables_, i.e. it is expected that they'll be defined in terms of other (non-computed) globals.  The other distinction is _scope_: globals are per-session whereas aliases usually refer to persistent data.

Reply:
Right. I think I'd feel better about this distinction if aliases weren't allowed to refer to global variables? That would be inconsistent with the idea that they are allowed anywhere, though?

I actually don't really love being able to have global variables anywhere, since it, like you said, makes it harder to identify references to global state, but allowing it would be nice in a bunch of cases.

Reply:
I'm OK with restricting globals to explicit queries and rewrite policy initially.